packages/mobile/src/components/audio/Audio.tsx

```diff
@@ -375,90 +373,117 @@ export const Audio = () => {
       uri: m3u8
-    }
-  }
-
-  let nextSource
-  if (nextOfflineTrackUri) {
-    nextSource = {
-      type: TrackType.Default,
-      uri: nextOfflineTrackUri
-    }
-    // TODO: remove feature flag - https://github.com/AudiusProject/audius-client/pull/2147
-  } else if (isStreamMp3Enabled && nextStreamingUri) {
-    nextSource = {
-      type: TrackType.Default,
-      uri: nextStreamingUri
-    }
-  } else if (nextM3u8) {
-    nextSource = {
+    }),
+    [m3u8]
+  )
+  const source = offlineSource ?? streamingSource ?? hlsSource
+
+  const nextOfflineSource = useMemo(
+    () =>
+      nextOfflineTrackUri
+        ? {
+            type: TrackType.Default,
+            uri: nextOfflineTrackUri
+          }
+        : null,
+    [nextOfflineTrackUri]
+  )
+  const nextStreamingSource = useMemo(
+    () =>
+      isStreamMp3Enabled && nextStreamingUri
+        ? {
+            type: TrackType.Default,
+            uri: nextStreamingUri
+          }
+        : null,
+    [isStreamMp3Enabled, nextStreamingUri]
+  )
+  const nextHlsSource = useMemo(
+    () => ({
       type: TrackType.HLS,
       uri: nextM3u8
-    }
-  }
+    }),
+    [nextM3u8]
+  )
+  const nextSource = nextOfflineSource ?? nextStreamingSource ?? nextHlsSource
 
   const currentUriRef = useRef<string | null>(null)
   const isPodcastRef = useRef<boolean>(false)
 
   const handleSourceChange = useCallback(async () => {
     const newUri = source.uri
-    if (currentUriRef.current !== newUri) {
-      currentUriRef.current = newUri
-      const imageUrl = trackImageSource?.source?.[2]?.uri ?? DEFAULT_IMAGE_URL
-      const nextImageUrl =
-        nextTrackImageSource?.source?.[2]?.uri ?? DEFAULT_IMAGE_URL
-
-      await TrackPlayer.reset()
-      // NOTE: Adding two tracks into the queue to make sure that android has a next button on the lock screen and notification controls
-      // This should be removed when the track player queue is used properly
-      await TrackPlayer.add([
-        {
-          url: newUri,
-          type: source.type ?? TrackType.Default,
-          title: track?.title,
-          artist: trackOwner?.name,
-          genre: track?.genre,
-          date: track?.created_at,
-          artwork: imageUrl,
-          duration: track?.duration
-        },
-        {
-          url: nextSource.uri,
-          type: nextSource.type ?? TrackType.Default,
-          title: nextTrack?.title,
-          artist: nextTrackOwner?.name,
-          genre: nextTrack?.genre,
-          date: nextTrack?.created_at,
-          artwork: nextImageUrl,
-          duration: nextTrack?.duration
-        }
-      ])
-
-      if (playing) await TrackPlayer.play()
-
-      const isPodcast = track?.genre === Genre.PODCASTS
-      if (isPodcast !== isPodcastRef.current) {
-        isPodcastRef.current = isPodcast
-        await updatePlayerOptions(isPodcast)
+    if (
+      loadingOfflineTrack ||
+      loadingNextOfflineTrack ||
+      currentUriRef.current === newUri
```

Comment:
Would this block non-offline tracks from playing if the offline track is still loading? Not sure if thats how it works, but wanted to check

Reply:
Yes. I don't think that should take much time, but we can do some perf tests. This was the only way I was able to get it to work, but I'm open to suggestions. Certainly don't want to impose any significant delay on all plays. Maybe we can check the offline downloads state and short circuit if it's not in there

Reply:
To clarify, it doesn't need to read the whole file, just verify that the file at the specified location exists

Reply:
sounds good. i'm gonna checkout the branch and test playback a bit just to be safe, but this seems ok to me

Reply:
Thank you! I actually think the short circuit may be worth it anyway. I'll look into adding it tonight